plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitSource.java

```diff
@@ -0,0 +1,196 @@
+import io.trino.plugin.jdbc.JdbcColumnHandle;
+import io.trino.plugin.jdbc.JdbcTableHandle;
+import io.trino.plugin.jdbc.PreparedQuery;
+import io.trino.plugin.jdbc.QueryBuilder;
+import io.trino.plugin.jdbc.logging.RemoteQueryModifier;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorSplit;
+import io.trino.spi.connector.ConnectorSplitSource;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.units.Duration.nanosSince;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_FILESYSTEM_ERROR;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_S3_CROSS_REGION_UNSUPPORTED;
+import static java.util.Objects.requireNonNull;
+
+public class RedshiftUnloadSplitSource
+        implements ConnectorSplitSource
+{
+    private static final Logger log = Logger.get(RedshiftUnloadSplitSource.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get();
+
+    private final JdbcClient jdbcClient;
+    private final QueryBuilder queryBuilder;
+    private final RemoteQueryModifier queryModifier;
+    private final Optional<String> unloadAuthorization;
+    private final String unloadOutputPath;
+    private final TrinoFileSystem fileSystem;
+    private final CompletableFuture<Void> resultSetFuture;
+
+    private boolean finished;
+
+    public RedshiftUnloadSplitSource(
+            ExecutorService executor,
+            ConnectorSession session,
+            JdbcClient jdbcClient,
+            JdbcTableHandle jdbcTableHandle,
+            List<JdbcColumnHandle> columns,
+            QueryBuilder queryBuilder,
+            RemoteQueryModifier queryModifier,
+            String unloadLocation,
+            Optional<String> unloadAuthorization,
+            TrinoFileSystem fileSystem)
+    {
+        requireNonNull(executor, "executor is null");
+        requireNonNull(session, "session is null");
+        this.jdbcClient = requireNonNull(jdbcClient, "jdbcClient is null");
+        requireNonNull(jdbcTableHandle, "jdbcTableHandle is null");
+        requireNonNull(columns, "columns is null");
+        this.queryBuilder = requireNonNull(queryBuilder, "queryBuilder is null");
+        this.queryModifier = requireNonNull(queryModifier, "queryModifier is null");
+        this.unloadAuthorization = requireNonNull(unloadAuthorization, "unloadAuthorization is null");
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+
+        String queryFragmentId = session.getQueryId() + "/" + UUID.randomUUID();
```

Comment:
Can you pls describe in which situation can we have multiple fragments?
I'm wondering why isn't it enough to rely on `session.getQueryId()` for having a unique place to unload the content.

Reply:
If we decide to rely only on `session.getQueryId()`, then we'd be able to add in `TestRedshiftUnload` also a test assertion to verify that after executing the query, the content of the unlod prefix corresponding to the queryId does not contain any file.

Reply:
Below query generates multiple fragments. 
```sql
use redshift.tpch;

select r.name,
       n1.name,
       n2.comment
from   nation n1
       join (select n_name      as name,
                    n_comment   as comment,
                    n_regionkey as regionkey
             from   tpch.tiny.nation) n2
         on n1.name = n2.name
       join (select *
             from   region) r
         on r.regionkey = n2.regionkey;
```

>If we decide to rely only on session.getQueryId(), then we'd be able to add in TestRedshiftUnload also a test assertion to verify that after executing the query, the content of the unlod prefix corresponding to the queryId does not contain any file.

We already check that queryId directory is empty at https://github.com/trinodb/trino/pull/24117/files#diff-07d07afd0a21f3a6c1de3a1bfc668fae6224e8698ba352d5dabfd876ab5f078dR105-R106

Reply:
I got it now (I believe).


Here is a simple version of the query:
```
explain select * from nation join (select * from region) r  on nation.regionkey = r.regionkey;

 Trino version: testversion                                                                                              >
 Fragment 0 [HASH]                                                                                                       >
     Output layout: [nationkey, name, regionkey, comment, regionkey, name_1, comment_2]                                  >
     Output partitioning: SINGLE []                                                                                      >
     Output[columnNames = [nationkey, name, regionkey, comment, regionkey, name, comment]]                               >
     │   Layout: [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152), regionkey:bigint, name_1:va>
     │   Estimates: {rows: 25 (5.25kB), cpu: 0, memory: 0B, network: 0B}                                                 >
     │   name := name_1                                                                                                  >
     │   comment := comment_2                                                                                            >
     └─ InnerJoin[criteria = (regionkey = regionkey_0), distribution = PARTITIONED]                                      >
        │   Layout: [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152), name_1:varchar(25), comm>
        │   Estimates: {rows: 25 (5.03kB), cpu: 8.36k, memory: 490B, network: 0B}                                        >
        │   Distribution: PARTITIONED                                                                                    >
        │   dynamicFilterAssignments = {regionkey_0 -> #df_234}                                                          >
        ├─ RemoteSource[sourceFragmentIds = [1]]                                                                         >
        │      Layout: [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]                      >
        └─ LocalExchange[partitioning = SINGLE]                                                                          >
           │   Layout: [regionkey_0:bigint, name_1:varchar(25), comment_2:varchar(152)]                                  >
           │   Estimates: {rows: 5 (490B), cpu: 0, memory: 0B, network: 0B}                                              >
           └─ RemoteSource[sourceFragmentIds = [2]]                                                                      >
                  Layout: [regionkey_0:bigint, name_1:varchar(25), comment_2:varchar(152)]                               >
                                                                                                                         >
 Fragment 1 [SOURCE]                                                                                                     >
     Output layout: [nationkey, name, regionkey, comment]                                                                >
     Output partitioning: HASH [regionkey]                                                                               >
     ScanFilter[table = redshift:tpch.nation testdb.tpch.nation, dynamicFilters = {regionkey = #df_234}]                 >
         Layout: [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]                            >
         Estimates: {rows: 25 (2.86kB), cpu: 2.86k, memory: 0B, network: 0B}/{rows: 25 (2.86kB), cpu: 2.86k, memory: 0B, >
         nationkey := nationkey:bigint:int8                                                                              >
         name := name:varchar(25):varchar                                                                                >
         comment := comment:varchar(152):varchar                                                                         >
         regionkey := regionkey:bigint:int8                                                                              >
                                                                                                                         >
 Fragment 2 [SOURCE]                                                                                                     >
     Output layout: [regionkey_0, name_1, comment_2]                                                                     >
     Output partitioning: HASH [regionkey_0]                                                                             >
     TableScan[table = redshift:tpch.region testdb.tpch.region]                                                          >
         Layout: [regionkey_0:bigint, name_1:varchar(25), comment_2:varchar(152)]                                        >
         Estimates: {rows: 5 (490B), cpu: 490, memory: 0B, network: 0B}                                                  >
         comment_2 := comment:varchar(152):varchar                                                                       >
         regionkey_0 := regionkey:bigint:int8                                                                            >
         name_1 := name:varchar(25):varchar                                                                              >
                                                                                                                         >
                                                                                                                         >
(1 row)

```

This makes sense in the light of https://trino.io/docs/current/sql/explain.html documentation.

Each fragment is basically a different query and therefore we need different fragment identifiers.

I observed in `io.trino.sql.planner.SplitSourceFactory.Visitor#visitTableScan` that we don't pass the `node.id` which would be the fragment identifier - but given that the data is removed right after running the query, we don't necessarily need to assign it the fragment identifier. At best, this could be considered as a follow-up.

Could you pls add an UNLOAD test involving 2 source fragments ? Just for avoiding regressions.